src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp

```diff
@@ -68,36 +68,46 @@ class NativeCallStackStorage : public CHeapObj<mtNMT> {
   };
 
 private:
-  struct Link : public ArenaObj {
-    Link* next;
+  struct Link;
+  using Allocator = HomogenousObjectArray<Link, mtNMT>;
+  using LinkPtr = typename Allocator::I;
+  LinkPtr nil() { return Allocator::nil; }
```

Comment:
could we have some clearer names?

- Link -> TableEntry or Entry (since its entries of the hashtable)
- LinkPtr -> TableEntryIndex

Also, could you put all data members of NCCS here, please? It makes it easier to see the implementation. The big boys here are _table, hash table entry storage, and stack storage.

I don't think you need a nil(), at least not if you keep referring to Allocator::nil elsewhere. 

Btw, why do you prefer using over  typedef?

Reply:
All suggestions sounds fine 

> Btw, why do you prefer using over typedef?

It's helpful that the name sticks out a lot more than for typedef when doing stuff like function pointer or array aliases, the name is always left of `=`. `using` also supports templates (you can do `template <typename T> using x = array<T>`).

That's why you might prefer `using` to `typedef`. I prefer it because that's what I'm used to using, boring but true.